Libraries/libutil/Sources/DefaultFilesystem.cpp

```diff
@@ -51,6 +52,168 @@ isExecutable(std::string const &path) const
     return ::access(path.c_str(), X_OK) == 0;
 }
 
+static Permissions
+ModePermissions(mode_t mode)
+{
+    Permissions permissions;
+    permissions.flag(Permissions::Flag::Sticky, (mode & S_ISVTX) != 0);
+    permissions.flag(Permissions::Flag::SetUserID, (mode & S_ISUID) != 0);
+    permissions.flag(Permissions::Flag::SetGroupID, (mode & S_ISGID) != 0);
+    permissions.user(Permissions::Permission::Read, (mode & S_IRUSR) != 0);
+    permissions.user(Permissions::Permission::Write, (mode & S_IWUSR) != 0);
+    permissions.user(Permissions::Permission::Execute, (mode & S_IXUSR) != 0);
+    permissions.group(Permissions::Permission::Read, (mode & S_IRGRP) != 0);
+    permissions.group(Permissions::Permission::Write, (mode & S_IWGRP) != 0);
+    permissions.group(Permissions::Permission::Execute, (mode & S_IXGRP) != 0);
+    permissions.other(Permissions::Permission::Read, (mode & S_IROTH) != 0);
+    permissions.other(Permissions::Permission::Write, (mode & S_IWOTH) != 0);
+    permissions.other(Permissions::Permission::Execute, (mode & S_IXOTH) != 0);
+    return permissions;
+}
+
+ext::optional<Permissions> DefaultFilesystem::
+readFilePermissions(std::string const &path) const
+{
+    struct stat st;
+    if (::stat(path.c_str(), &st) < 0) {
+        return ext::nullopt;
+    }
+
+    return ModePermissions(st.st_mode);
+}
+
+ext::optional<Permissions> DefaultFilesystem::
+readSymbolicLinkPermissions(std::string const &path) const
+{
+    struct stat st;
+    if (::lstat(path.c_str(), &st) < 0) {
+        return ext::nullopt;
+    }
+
+    return ModePermissions(st.st_mode);
+}
+
+ext::optional<Permissions> DefaultFilesystem::
+readDirectoryPermissions(std::string const &path) const
```

Comment:
I'm not quite sure what to do with these APIs. Ideally, there'd just be one read and write permissions API. But since (read+write) symlink permissions are non-portable and (write only) directory permissions need an extra `bool recursive` parameter, I'm not sure what's better than six different functions to be super clear. O_o

In terms of implementation, I think it'd save like one line to make a function and avoid copying the 4 lines inside? Not sure if the savings is worth the indirection...